Source/WebCore/dom/Element.cpp

```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
+
+        // Specs: If element is an input element whose type attribute is in the Telephone state
+        // then return LTR;
+        if (auto* input = dynamicDowncast<HTMLInputElement>(element); input && input->isTelephoneField())
+            return TextDirection::LTR;
+
+        // Specs: If the parent is a shadow root, then return the directionality of parent's host.
+        // Specs: If the parent is an element, then return the directionality of parent.
+        if (RefPtr parent = element.parentOrShadowHostElement())
+            return computeElementTextDirection(*parent, elementTextDirectionState(*parent));
+
+        return std::nullopt;
     }
 
-    if (parent && parent->selfOrPrecedingNodesAffectDirAuto()) {
-        if (isValid && direction != TextDirectionDirective::Auto)
-            setHasDirAutoFlagRecursively(this, false);
-        parent->adjustDirectionalityIfNeededAfterChildAttributeChanged(this);
+    return std::nullopt;
+}
+
+static std::optional<TextDirection> computeElementContainedTextAutoDirection(const Element& element)
+{
+    // Specs: https://html.spec.whatwg.org/multipage/dom.html#contained-text-auto-directionality
+    for (RefPtr child = element.firstChild(); child; ) {
+        // Specs: Skip bdi, script, style nodes.
+        if (child->hasTagName(bdiTag) || child->hasTagName(scriptTag) || child->hasTagName(styleTag)) {
+            child = NodeTraversal::nextSkippingChildren(*child, &element);
+            continue;
+        }
+
+        if (auto* childElement = dynamicDowncast<Element>(*child)) {
+            // Specs: Skip text form controls
+            // Specs: Skip the element whose dir attribute is not in the undefined state
+            if (childElement->isTextField() || elementHasValidTextDirectionState(*childElement)) {
+                child = NodeTraversal::nextSkippingChildren(*child, &element);
+                continue;
+            }
+        }
+
+        // Specs: If child is a slot element whose root is a shadow root,
+        // then return the directionality of that shadow root's host.
+        if (auto* childSlotElement = dynamicDowncast<HTMLSlotElement>(*child)) {
+            if (auto childHost = childSlotElement->shadowHost())
+                return computeElementTextDirection(*childHost, elementTextDirectionState(*childHost));
+        }
+
+        // Specs: If descendant is a Text node and its text direction is not null,
+        // then return it.
+        if (child->isTextNode()) {
+            if (auto direction = computeTextDirectionFromText(child->textContent(true)))
+                return direction;
+        }
+
+        child = NodeTraversal::next(*child, &element);
     }
+
+    return std::nullopt;
 }
 
-void Element::updateEffectiveDirectionality(std::optional<TextDirection> direction)
+static std::optional<TextDirection> computeSlotElementTextDirection(const HTMLSlotElement& slotElement)
```

Comment:
I'd rather call this `computeTextDirectionOfSlotElement`.

Reply:
Done.